apis/Google.Cloud.Firestore/Google.Cloud.Firestore/ValueComparer.cs

```diff
@@ -0,0 +1,251 @@
+        }
+
+        private int CompareTimestamps(Value left, Value right)
+        {
+            var leftTimestamp = left.TimestampValue;
+            var rightTimestamp = right.TimestampValue;
+            int result = leftTimestamp.Seconds.CompareTo(rightTimestamp.Seconds);
+            return result != 0 ? result : leftTimestamp.Nanos.CompareTo(rightTimestamp.Nanos);
+        }
+
+        private int CompareGeoPoints(Value left, Value right)
+        {
+            var leftPoint = left.GeoPointValue;
+            var rightPoint = right.GeoPointValue;
+            int result = leftPoint.Latitude.CompareTo(rightPoint.Latitude);
+            return result != 0 ? result : leftPoint.Longitude.CompareTo(rightPoint.Longitude);
+        }
+
+        private int CompareResourcePaths(Value left, Value right)
+        {
+            // Note: this does no validation, but it's simple...
+            // We can't just do a string comparison, because of cases such as:
+            // foo/bar/baz
+            // foo/bar-/baz
+            // Here "bar-" shoudl come after "bar" because it's longer, but '-' is earlier than '/'.
+            string leftReference = left.ReferenceValue;
+            string rightReference = right.ReferenceValue;
+            int length = Math.Min(leftReference.Length, rightReference.Length);
+            for (int i = 0; i < length; i++)
+            {
+                char leftChar = leftReference[i];
+                char rightChar = rightReference[i];
+                if (leftChar != rightChar)
+                {
+                    return leftChar == '/' ? -1 // Left segment finishes earlier
+                        : rightChar == '/' ? 1  // Right segment finishes earlier
+                        : leftChar - rightChar; // Not a segment end, so just compare ordinals
+                }
+            }
+            // Shorter comes before longer
+            return leftReference.Length.CompareTo(rightReference.Length);
+        }
+
+        private int CompareArrays(Value left, Value right)
+        {
+            var leftArray = left.ArrayValue.Values;
+            var rightArray = right.ArrayValue.Values;
+            
+            int size = Math.Min(leftArray.Count, rightArray.Count);
+            for (int i = 0; i < size; i++)
+            {
+                int result = Compare(leftArray[i], rightArray[i]);
+                if (result != 0)
+                {
+                    return result;
+                }
+            }
+            return leftArray.Count.CompareTo(rightArray.Count);
+        }
+
+        private int CompareObjects(Value left, Value right)
+        {
+            // This requires iterating over the keys in the object in order and doing a
+            // deep comparison.
+            var leftMap = new SortedDictionary<string, Value>(left.MapValue.Fields, StringComparer.Ordinal);
+            var rightMap = new SortedDictionary<string, Value>(right.MapValue.Fields, StringComparer.Ordinal);
```

Comment:
I don't think `SortedDictionary`'s are required.
Could you use something like `left.MapValue.Fields.OrderBy(x => x.Key)`